cpp/src/IceGrid/NodeCache.cpp

```diff
@@ -840,29 +790,54 @@ NodeEntry::finishedRegistration()
     if(_registering)
     {
         _registering = false;
-        notifyAll();
+        _condVar.notify_all();
     }
 }
 
 void
-NodeEntry::finishedRegistration(const Ice::Exception& ex)
+NodeEntry::finishedRegistration(exception_ptr exptr)
 {
-    Lock sync(*this);
+    lock_guard lock(_mutex);
     if(_cache.getTraceLevels() && _cache.getTraceLevels()->node > 0)
     {
-        Ice::Trace out(_cache.getTraceLevels()->logger, _cache.getTraceLevels()->nodeCat);
-        out << "node `" << _name << "' session creation failed:\n" << ex;
+        try
+        {
+            rethrow_exception(exptr);
+        }
+        catch(const std::exception& ex)
+        {
+            Ice::Trace out(_cache.getTraceLevels()->logger, _cache.getTraceLevels()->nodeCat);
+            out << "node `" << _name << "' session creation failed:\n" << ex.what();
+        }
     }
 
     if(_registering)
     {
         _registering = false;
-        notifyAll();
+        _condVar.notify_all();
     }
 }
 
-ServerDescriptorPtr
-NodeEntry::getServerDescriptor(const ServerInfo& server, const SessionIPtr& session)
+// Return the weak_ptr's copy or, if nullptr, create a new "self removing" shared_ptr NodeEntry from 'this'
+// which removes itself from the NodeCache upon destruction
+shared_ptr<NodeEntry>
+NodeEntry::selfRemovingPtr() const
+{
+    lock_guard lock(_selfRemovingMutex);
+
+    auto entry = _selfRemovingPtr.lock();
+
+    if (!entry)
+    {
+        entry = shared_ptr<NodeEntry>(const_cast<NodeEntry*>(this), [&](NodeEntry*) { _cache.remove(_name); });
```

Comment:
It would be cleaner to use the NodeEntry* parameter over capturing this and using its data members.